### PR TITLE
Booting guest with kernel param testcase

### PR DESCRIPTION
**Pull request by balamuruhans**
1. edit_kernel_param_only can be used to change the kernel param
inside the guest image and proceed with other tests, similar to
guest_import for exercising the testcase with different kernel param options
for all archs
    - add_kernel_param: adds the kernel param and reboots the guest
    - remove_kernel_param: removes the kernel param and reboots the guest

2. Added testcases for bringing up POWER9 guest in radix or hpt MMU based on
the kernel param for PowerPC arch.

Signed-off-by: Balamuruhan S <bala24@linux.vnet.ibm.com>
Signed-off-by: Satheesh Rajendran <sathnaga@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by balamuruhans** [comment]
Failure is known issue and will be fixed in latest kernel.

```
JOB ID     : 7296c99cf9c49f24256fac973b783e1270d84438
JOB LOG    : /home/bala/avocado-fvt-wrapper/results/job-2017-06-29T11.06-7296c99/job.log
 (1/4) guest_boot.qemu.qcow2.virtio_scsi.smp2.virtio_net.Fedora.26.ppc64le.powerkvm-libvirt.vm_boot_with_kernel_param.edit_kernel_param_only.add_kernel_param: PASS (98.05 s)
 (2/4) guest_boot.qemu.qcow2.virtio_scsi.smp2.virtio_net.Fedora.26.ppc64le.powerkvm-libvirt.vm_boot_with_kernel_param.edit_kernel_param_only.remove_kernel_param: PASS (101.29 s)
 (3/4) guest_boot.qemu.qcow2.virtio_scsi.smp2.virtio_net.Fedora.26.ppc64le.powerkvm-libvirt.vm_boot_with_kernel_param.boot_guest.hpt_guest: FAIL (98.94 s)
 (4/4) guest_boot.qemu.qcow2.virtio_scsi.smp2.virtio_net.Fedora.26.ppc64le.powerkvm-libvirt.vm_boot_with_kernel_param.boot_guest.radix_guest: PASS (99.33 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 404.37 s
JOB HTML   : /home/bala/avocado-fvt-wrapper/results/job-2017-06-29T11.06-7296c99/html/results.html

```

File: libvirt/tests/src/vm_boot_with_kernel_param.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                        test.fail("Able to find %s in dmesg of guest: "
+                                  "%s" % (boot_log, vm.name))
+                    logging.info("unable to find %s in dmesg of guest: %s",
+                                 boot_log, vm.name)
+                else:
+                    if boot_log not in output:
+                        test.fail("unable to find %s in dmesg of guest: "
+                                  "%s" % (boot_log, vm.name))
+                    logging.info("Able to find %s in dmesg of guest: %s",
+                                 boot_log, vm.name)
```
Comment (sathnaga) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: session can be closed here inside the loop, failure path let is handle by `finally`

File: libvirt/tests/src/vm_boot_with_kernel_param.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    kernel_param = params.get("kernel_param", "quiet")
+    kernel_param_remove = params.get("kernel_param_remove", "")
+    if not kernel_param:
+        kernel_param = None
+    if not kernel_param_remove:
+        kernel_param_remove = None
+    cpu_check = params.get("hardware", "").upper()
+    boot_log = params.get("boot_log", None)
+    status_error = params.get("status_error", "no") == "yes"
+    vm_dict = {}
```
Comment (sathnaga) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: I guess this vm_dict can be avoided as we do not change any guest xml related stuff, same on further usage.

File: libvirt/tests/src/vm_boot_with_kernel_param.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        cpu_model = utils_misc.get_cpu_info()['Model name'].upper()
+        if cpu_check not in cpu_model:
+            logging.info("This test will work for %s", cpu_check)
+            test.skip("Test is not applicable for %s" % cpu_model)
+    # back up vmxml
+    for vm_name in vms:
+        vm_list.append(env.get_vm(vm_name))
+        vm_dict[vm_name] = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    try:
+        for vm in vm_list:
```
Comment (sathnaga) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: ```for vm in env.get_all_vms():``` can be used instead

File: libvirt/tests/src/vm_boot_with_kernel_param.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                    cmd = "grep cpu /proc/cpuinfo | awk '{print $3}' | "
+                    cmd += "head -n 1"
+                    status, output = session.cmd_status_output(cmd)
+                    if status:
+                        test.error("couldn't get cpu information from guest "
+                                   "%s" % vm.name)
+                    if cpu_check not in output.upper() and "radix" in boot_log:
+                        test.skip("radix MMU not supported in %s" % output)
+                cmd = "dmesg"
+                status, output = session.cmd_status_output(cmd)
```
Comment (sathnaga) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: 'dmesg' can be directly given as single use

File: libvirt/tests/src/vm_boot_with_kernel_param.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                        test.fail("unable to find %s in dmesg of guest: "
+                                  "%s" % (boot_log, vm.name))
+                    logging.info("Able to find %s in dmesg of guest: %s",
+                                 boot_log, vm.name)
+    finally:
+        # close the session and recover the vms
+        if session:
+            session.close()
+        for vm in vm_list:
+            vm.destroy()
```
Comment (sathnaga) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: am not sure if destroy is required, given that kill_vm* is set in params, checkout and can be removed

File: libvirt/tests/src/vm_boot_with_kernel_param.py (hunk 2 above)
Comment (balamuruhans) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: This one, I need it for future work, so lets have it.

File: libvirt/tests/src/vm_boot_with_kernel_param.py (hunk 3 above)
Comment (balamuruhans) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: Sure, I will use it.

File: libvirt/tests/src/vm_boot_with_kernel_param.py (hunk 4 above)
Comment (balamuruhans) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: sure.

File: libvirt/tests/src/vm_boot_with_kernel_param.py (hunk 1 above)
Comment (balamuruhans) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: sure

File: libvirt/tests/src/vm_boot_with_kernel_param.py (hunk 5 above)
Comment (balamuruhans) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: I can see kill_vm param kills only one VM in the environment in postprocess

File: libvirt/tests/src/vm_boot_with_kernel_param.py (hunk 3 above)
Comment (balamuruhans) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: we have to use env.get_all_vms() twice here and in clean up, so I will save it in a variable vm_list and proceed.

**Comment by balamuruhans** [comment]
@sathnaga, I have made the changes as per your suggestion.

File: libvirt/tests/src/vm_boot_with_kernel_param.py (hunk 5 above)
Comment (sathnaga) [inline libvirt/tests/src/vm_boot_with_kernel_param.py]: can be a separate PR taken care in postprocess vm @balamuruhans
